aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -499,8 +521,350 @@ Tensor coo_to_sparse_csr(const Tensor& self) {
 }
 
+Tensor coo_to_sparse_csc(const Tensor& self) {
+  AT_ERROR(
+      "Conversion from ", self.layout(), " to CSC is currently not supported.");
+  return self;
+}
+
+Tensor coo_to_sparse_bsr(const Tensor& self, IntArrayRef blocksize) {
+  AT_ERROR(
+      "Conversion from ", self.layout(), " to BSR is currently not supported.");
+  return self;
+}
+
+Tensor coo_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
+  AT_ERROR(
+      "Conversion from ", self.layout(), " to BSC is currently not supported.");
+  return self;
+}
+
+namespace {
+template <typename input_t, typename output_t>
+void convert_indices_from_coo_to_csr_cpu(
+    const Tensor& result,
+    const Tensor& input,
+    const int64_t size) {
+  int64_t numel = input.numel();
+  const input_t* data_in = input.data_ptr<input_t>();
+  output_t* data_out = result.data_ptr<output_t>();
+
+  if (numel == 0) {
+    result.zero_();
+    return;
+  }
+
+  for (int64_t i = 0; i <= data_in[0]; i++)
+    data_out[i] = static_cast<output_t>(0);
+
+  at::parallel_for(
+      0, numel - 1, at::internal::GRAIN_SIZE, [&](int64_t start, int64_t end) {
+        input_t curr_value = data_in[start], next_value;
+        for (const auto i : c10::irange(start, end)) {
+          next_value = data_in[i + 1];
+          for (; curr_value < next_value; curr_value++)
+            data_out[curr_value + 1] = static_cast<output_t>(i + 1);
+        }
+      });
+  for (int64_t i = data_in[numel - 1] + 1; i < size + 1; i++) {
+    data_out[i] = static_cast<output_t>(numel);
+  }
+}
+
+template <typename input_t, typename output_t>
+void convert_indices_from_csr_to_coo_cpu(
+    const Tensor& indices,
+    const Tensor& crow_indices,
+    const Tensor& col_indices,
+    const bool transpose = false) {
+  int64_t nrows = crow_indices.numel() - 1;
+  if (nrows == 0) {
+    indices.zero_();
+    return;
+  }
+  auto crow_indices_ = crow_indices.expect_contiguous();
+  const input_t* crow_indices_data_in = crow_indices_->data_ptr<input_t>();
+  TORCH_INTERNAL_ASSERT(indices.is_contiguous());
+  auto row0 = indices.select(0, transpose ? 1 : 0);
+  auto row1 = indices.select(0, transpose ? 0 : 1);
+  output_t* data_out = row0.data_ptr<output_t>();
+  row1.copy_(*col_indices.expect_contiguous());
+  at::parallel_for(
+      0, nrows, at::internal::GRAIN_SIZE, [&](int64_t start, int64_t end) {
+        for (const auto i : c10::irange(start, end)) {
+          std::fill(
+              &data_out[crow_indices_data_in[i]],
+              &data_out[crow_indices_data_in[i + 1]],
+              static_cast<output_t>(i));
+        }
+      });
+}
+} // namespace
+
+TORCH_IMPL_FUNC(_convert_indices_from_coo_to_csr_structured_cpu)
+(const Tensor& input,
+ const int64_t size,
+ const bool out_int32,
+ const Tensor& result) {
+  if (out_int32) {
+    AT_DISPATCH_INTEGRAL_TYPES(
+        input.scalar_type(), "convert_indices_from_coo_to_csr_cpu", [&] {
+          convert_indices_from_coo_to_csr_cpu<scalar_t, int>(
```

Comment:
Nit: `int` may be 16-bit, so
```suggestion
          convert_indices_from_coo_to_csr_cpu<scalar_t, int32_t>(
```